src/plugins/ui_stateful/edition.ts

```diff
@@ -565,18 +537,54 @@ export class EditionPlugin extends UIPlugin {
     const ref = this.getZoneReference(zone);
     if (this.canStartComposerRangeSelection()) {
       this.insertText(ref, start);
-      this.selectionInitialStart = start;
     } else {
       this.insertText("," + ref, start);
-      this.selectionInitialStart = start + 1;
     }
   }
+
   /**
    * Replace the current reference selected by the new one.
    * */
-  private replaceSelectedRanges(zone: Zone | UnboundedZone) {
+  private replaceSelectedRange(zone: Zone | UnboundedZone) {
     const ref = this.getZoneReference(zone);
-    this.replaceText(ref, this.selectionInitialStart, this.selectionEnd);
+    const currentToken = this.getTokenAtCursor();
+    const start = currentToken?.type === "REFERENCE" ? currentToken.start : this.selectionStart;
+    this.replaceText(ref, start, this.selectionEnd);
+  }
+
+  /**
+   * Replace the reference of the old zone by the new one.
+   */
+  private updateComposerRange(oldZone: Zone, newZone: Zone | UnboundedZone) {
+    const activeSheetId = this.getters.getActiveSheetId();
+
+    const tokentAtCursor = this.getTokenAtCursor();
+    const tokens = tokentAtCursor ? [tokentAtCursor, ...this.currentTokens] : this.currentTokens;
+    const previousRefToken = tokens
+      .filter((token) => token.type === "REFERENCE")
+      .find((token) => {
+        const { xc, sheetName: sheet } = splitReference(token.value);
+        const sheetName = sheet || this.getters.getSheetName(this.sheetId);
+
+        if (this.getters.getSheetName(activeSheetId) !== sheetName) {
+          return false;
+        }
+        const refRange = this.getters.getRangeFromSheetXC(activeSheetId, xc);
+        return isEqual(this.getters.expandZone(activeSheetId, refRange.zone), oldZone);
```

Comment:
Maybe a helper/getter for this matcher ? We have the exact same code in `selection_input.ts`

Reply:
I don't really see what we could generalize more than that, something like "isZoneEqual(sheetId, zone1, zone2) ? not sure it's worth it tbh